src/Language/PureScript/Graph.hs

```diff
@@ -0,0 +1,64 @@
+module Language.PureScript.Graph (graph) where
+
+import Prelude.Compat
+
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.Map as Map
+import qualified Data.Aeson as Json
+
+import           Control.Monad (forM, foldM)
+import           Data.Aeson ((.=))
+import           Data.Map (Map)
+import           Data.ByteString.Lazy.UTF8 (ByteString)
+import           Data.Text (Text)
+import           System.IO.UTF8 (readUTF8FileT)
+
+import qualified Language.PureScript.CST as CST
+import qualified Language.PureScript.Make as Make
+import qualified Language.PureScript.ModuleDependencies as Dependencies
+import qualified Language.PureScript.Options as Options
+
+import           Language.PureScript.Errors (MultipleErrors)
+import           Language.PureScript.Names (ModuleName, runModuleName)
+
+
+graph :: [FilePath] -> IO (MultipleErrors, Either MultipleErrors ByteString)
+graph input = do
+  moduleFiles <- readInput input
+  (makeResult, makeWarnings) <- Make.runMake Options.defaultOptions $ do
+    ms <- CST.parseModulesFromFiles id moduleFiles
+    let parsedModuleSig = Dependencies.moduleSignature . CST.resPartial
+    (_sorted, moduleGraph) <- Dependencies.sortModules (parsedModuleSig . snd) ms
+    let swap (a, b) = (b, a)
+    let pathMap = Map.fromList
+                . fmap (swap . fmap (Dependencies.sigModuleName . parsedModuleSig))
+                $ ms
+    pure (moduleGraphToJSON pathMap moduleGraph)
+  pure $ case makeResult of
+    Left makeErrors -> (makeWarnings, Left makeErrors)
+    Right Nothing -> undefined -- TODO: what should happen here?
+    Right (Just json) -> (makeWarnings, Right $ Json.encode json)
+
+
+moduleGraphToJSON
+  :: Map ModuleName FilePath -> Dependencies.ModuleGraph -> Maybe Json.Value
+moduleGraphToJSON paths = fmap Json.Object . foldM insert mempty
```

Comment:
The `Maybe` that's returned here is the case I talked about above. If we want to report it as an error this will need to become an `Either [ModuleName] Json.Value` that can hold the module names that we didn't end up finding (or even a NonEmptyList). This code can also simplified, the `key` binding can just be inlined and the mixture of `<$>`, `fmap` and do-notation to deal with `Maybe`'s can be unified. Something like:

```haskell
moduleGraphToJSON
  :: Map ModuleName FilePath
  -> Dependencies.ModuleGraph
  -> Maybe Json.Value
moduleGraphToJSON paths = fmap Json.Object . foldM insert mempty
  where
  insert :: Json.Object -> (ModuleName, [ModuleName]) -> Maybe Json.Object
  insert obj (mn, depends) = Map.lookup mn paths <&> \path ->
    let
      value = Json.object
        [ "path"  .= path
        , "depends" .= fmap runModuleName depends
        ]
    in HashMap.insert (runModuleName mn) value obj

```

Reply:
I would have expected that if a module imports another one which doesn't exist, it would be reported via the normal `MonadError` channel from `sortModules`, and we wouldn't reach this point in the first place. If that's correct, then I think we should write
```haskell
insert :: Json.Object -> (ModuleName, [ModuleName]) -> Json.Object
insert obj (mn, depends) = HashMap.insert key value obj
  where
  path = fromMaybe (P.internalError "missing module name in graph") $
    Map.lookup mn paths
  value = [...]
```

Reply:
(because in this case, a missing module name would indicate a bug in the compiler)

Reply:
I think you're right. Probably a good idea to verify though (maybe even add a test for that? So we have one passing, one failing?)

Reply:
Added a failing test that verifies that all the above is correct :+1: 
Thanks for the input!